github/commitstatus.go

```diff
@@ -172,3 +214,22 @@ func min(a, b int) int {
 	}
 	return b
 }
+
+// checkForRetry determines if we should retry the http request and caluclates wait time between retries
+func checkForRetry(res httpResponse, attempt int) (bool, time.Duration) {
+	switch {
+	case attempt == maxRetries-1:
```

Comment:
With the above loop modification:
```suggestion
	case attempt == maxRetries:
```

And now I see why you wondered about the `return nil` above.
I've not thought it trough but you can probably ignore that case here and let the loop terminate and handle the maxRetries as I described it above.

Reply:
but the sleep is after... so we need to avoid sleeping before exiting the loop, which means I would just have to move this case as if else in the loop... which I didn't like...

Reply:
> I've not thought it trough but you can probably ignore that case here and let the loop terminate and handle the maxRetries as I described it above.

See:
```
retry, timeToSleep := checkForRetry(resp, attempt)
if !retry {
   return s.checkStatus(resp, state, sha, url)
}
time.Sleep(timeToSleep)
```

In our discussion we said that `retry` should determine if we should retry or not. If max attempts is reached we should not retry. So this beahaviour is in accordance with what we agreed....

If on other hand I remove the case `case attempt == maxRetries:` we would say this is a retryable error, and we would reach: `time.Sleep(timeToSleep)`. Problem is that `timeToSleep` can be anything reported in the last request... so this is potentially problematic.. to avoid it I would need to introduce an if statement `if attempt == maxRetries` and return from this statement...
This is just reapeating `case attempt == maxRetries:` but in a wrong way...

Reply:
After discussion with @odormond in private chat this issue was addressed in:
https://github.com/Pix4D/cogito/pull/118/commits/c8f7ab2d1fbd2f8ea898ee590f2500d2b7d5fe71